### ISSUE TITLE
Spinning cars while following trajectories with velocity close to zero

### ISSUE THREAD

**Issue by mschneider1711** [esmini/esmini#199]
Hi,

I got an issue with spinning cars that use the FollowTrajectoryAction. It occurs when the vehicle has a very small velocity close to zero.  
Is this a known issue?

Thanks for the response!

**Comment by eknabevcc** [comment]
Please provide an example. It could be a complete scenario or just the portion where the trajectory is defined. And we'll have a look. 

I have some ideas but it depends, so it's better to look at an actual example to avoid confusion.

**Comment by mschneider1711** [comment]
Is it possible to send it by E-Mail to you? That would be great.

**Comment by eknabevcc** [comment]
The preferred way would be to post just a snipped here (for reference of the issue). E.g. the last portion of the FollowTrajectoryAction including the last two vertices would be enough. Like:
```
                <Vertex time="4">
                    <Position>
                        <RelativeLanePosition dLane="1" ds="110" entityRef="Target" offset="2"/>
                    </Position>
                </Vertex>
                <Vertex time="6">
                    <Position>
                        <RelativeLanePosition dLane="0" ds="165" entityRef="Target" offset="0"/>
                    </Position>
                </Vertex>
            </Polyline>
        </Shape>
    </Trajectory>
    <TimeReference>
        <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/>
    </TimeReference>
    <TrajectoryFollowingMode followingMode="position"/>
</FollowTrajectoryAction>

```
You can even replace or delete the actual position values. Just make sure the format is preserved. That would be very helpful, much better than nothing.
Let's start from there, and if we need to dig deeper we'll resort to mail.

**Comment by mschneider1711** [comment]
`<Action name="TrajectoryAction">
                                <PrivateAction>
                                    <RoutingAction>
                                        <FollowTrajectoryAction>
                                            <Trajectory name="EgoVehicle_Trajectory" closed="false">
                                                <ParameterDeclarations/>
                                                <Shape>
                                                    <Polyline>
                                                        <Vertex time="0.0">
                                                            <Position>
                                                                <WorldPosition x="42.79356" y="-22.41851" z="0" h="5.0434287212031075" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="1.0">
                                                            <Position>
                                                                <WorldPosition x="43.13015" y="-23.46412" z="0" h="5.025733875114689" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="2.0">
                                                            <Position>
                                                                <WorldPosition x="43.17322" y="-23.62474" z="0" h="5.026297267397232" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="3.0">
                                                            <Position>
                                                                <WorldPosition x="43.17815" y="-23.62238" z="0" h="5.022383890148411" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="4.0">
                                                            <Position>
                                                                <WorldPosition x="43.17577" y="-23.62334" z="0" h="5.023869688940633" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="5.0">
                                                            <Position>
                                                                <WorldPosition x="43.18089" y="-23.62193" z="0" h="5.024440586138961" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="6.0">
                                                            <Position>
                                                                <WorldPosition x="43.18062" y="-23.62596" z="0" h="5.025354789601155" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="7.0">
                                                            <Position>
                                                                <WorldPosition x="43.18295" y="-23.63462" z="0" h="5.023539647179081" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="8.0">
                                                            <Position>
                                                                <WorldPosition x="43.18903" y="-23.64465" z="0" h="5.020450763468902" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="9.0">
                                                            <Position>
                                                                <WorldPosition x="43.19063" y="-23.64605" z="0" h="5.020968079059193" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="10.0">
                                                            <Position>
                                                                <WorldPosition x="43.1857" y="-23.64504" z="0" h="5.019423288138253" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="11.0">
                                                            <Position>
                                                                <WorldPosition x="43.18164" y="-23.62882" z="0" h="5.022417051404199" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="12.0">
                                                            <Position>
                                                                <WorldPosition x="43.28292" y="-23.91932" z="0" h="5.023329684070066" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="13.0">
                                                            <Position>
                                                                <WorldPosition x="43.90458" y="-25.32657" z="0" h="5.104151867472645" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>`
                                                        
The issue takes place until around 13s in simulation time. Then the vehicle starts to follow the trajectory. Because of traffic regulations the car needs to let the other vehicles in the scene pass by and therefore has a very small velocity.

**Comment by mschneider1711** [comment]
> <Action name="TrajectoryAction">
                                <PrivateAction>
                                    <RoutingAction>
                                        <FollowTrajectoryAction>
                                            <Trajectory name="EgoVehicle_Trajectory" closed="false">
                                                <ParameterDeclarations/>
                                                <Shape>
                                                    <Polyline>
                                                        <Vertex time="0.0">
                                                            <Position>
                                                                <WorldPosition x="42.79356" y="-22.41851" z="0" h="5.0434287212031075" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="1.0">
                                                            <Position>
                                                                <WorldPosition x="43.13015" y="-23.46412" z="0" h="5.025733875114689" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="2.0">
                                                            <Position>
                                                                <WorldPosition x="43.17322" y="-23.62474" z="0" h="5.026297267397232" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="3.0">
                                                            <Position>
                                                                <WorldPosition x="43.17815" y="-23.62238" z="0" h="5.022383890148411" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="4.0">
                                                            <Position>
                                                                <WorldPosition x="43.17577" y="-23.62334" z="0" h="5.023869688940633" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="5.0">
                                                            <Position>
                                                                <WorldPosition x="43.18089" y="-23.62193" z="0" h="5.024440586138961" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="6.0">
                                                            <Position>
                                                                <WorldPosition x="43.18062" y="-23.62596" z="0" h="5.025354789601155" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="7.0">
                                                            <Position>
                                                                <WorldPosition x="43.18295" y="-23.63462" z="0" h="5.023539647179081" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="8.0">
                                                            <Position>
                                                                <WorldPosition x="43.18903" y="-23.64465" z="0" h="5.020450763468902" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="9.0">
                                                            <Position>
                                                                <WorldPosition x="43.19063" y="-23.64605" z="0" h="5.020968079059193" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="10.0">
                                                            <Position>
                                                                <WorldPosition x="43.1857" y="-23.64504" z="0" h="5.019423288138253" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="11.0">
                                                            <Position>
                                                                <WorldPosition x="43.18164" y="-23.62882" z="0" h="5.022417051404199" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="12.0">
                                                            <Position>
                                                                <WorldPosition x="43.28292" y="-23.91932" z="0" h="5.023329684070066" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="13.0">
                                                            <Position>
                                                                <WorldPosition x="43.90458" y="-25.32657" z="0" h="5.104151867472645" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>

**Comment by mschneider1711** [comment]
Because of traffic regulations, this vehicle has to let the other vehicles pass the intersection before itself. Therefore the velocity in this situation is very small. After approximately 12s in simulation time, the vehicle starts to follow the trajectory correctly.

**Comment by eknabevcc** [comment]
Great, thanks, this will help!
I just miss the very last portion where the TimeReference is specified. I assume you have not "None" but either "relative" or "absolute".  E.g:
```
<TimeReference>
    <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/>
</TimeReference>
```
Maybe you can just provide that part as well.

Then as a final information, what version of esmini do you run? Just so we can reproduce correctly. (version is found either in log file, or in file version.txt found in esmini root folder). 

BTW: Release 2.18 included a fix for Polyline based trajectories, so that any specified heading values is respected which was not the case previously. So it would be good to know if you're using a more recent one or not.

**Comment by mschneider1711** [comment]
> <Vertex time="21.0">
                                                            <Position>
                                                                <WorldPosition x="76.01878" y="-26.40582" z="0" h="0.20790152610246215" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="22.0">
                                                            <Position>
                                                                <WorldPosition x="82.56334" y="-25.04803" z="0" h="0.20904803288809723" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                        <Vertex time="23.0">
                                                            <Position>
                                                                <WorldPosition x="89.27692" y="-23.72944" z="0" h="0.20257757375217864" p="0" r="0"/>
                                                            </Position>
                                                        </Vertex>
                                                    </Polyline>
                                                </Shape>
                                            </Trajectory>
                                            <TimeReference>
                                                <Timing domainAbsoluteRelative="relative" scale="1" offset="0"/>
                                            </TimeReference>
                                            <TrajectoryFollowingMode followingMode="position"/>
                                        </FollowTrajectoryAction>
                                    </RoutingAction>
                                </PrivateAction>
                            </Action>

**Comment by mschneider1711** [comment]
I am using Esmini v2.16.1

**Comment by mschneider1711** [comment]
I just tested v2.18.2 and in this case, it works. 
Thank you for your help.

**Comment by eknabevcc** [comment]
Ah, Ok, I see. Then I think we the root cause is that the headings provided in the trajectory definition is ignored. Instead esmini is always calculating heading based on the X, Y motion (delta between vertices). 
A more correct behavior is to use whatever is provided and only calculate if no heading is provided. And that was implemented for release 2.18.0.

Ah, I see. You already found out. Thanks for reporting the issue and quick feedback!
